### PR TITLE
clean up and finish fixing #2782

### LINKED ISSUE #2782: Performance regression in dnu restore
Restore has a major perf regression. Prior to (and including) DNX 1.0.0-beta8-15641 Microsoft.AspNet.Mvc.FunctionalTests would restore in ~70 seconds, now it takes over 15mins.

I have a fix ready (just doing final correctness checks); I'm filing this to track the work. A PR will be coming for review, but I'll be merging it immediately to unblock builds.

/cc @dougbu @davidfowl @rynowak

### PR DESCRIPTION
**Pull request by anurse**
Fixes #2782 

This fixes a small correctness issue (that isn't affecting anything yet) in the original fix and does some clean-up to try and reduce the number of times we scan the runtime specs.

/cc @davidfowl @dougbu @rynowak

### CODE REVIEW

File: src/Microsoft.Dnx.Tooling/Restore/RestoreCommand.cs [hunk 1]
```diff
@@ -600,7 +606,16 @@ private IEnumerable<string> GetRestoreRuntimes(IEnumerable<string> projectRuntim
...
-                    effectiveRuntimeSpecs.Add(runtimeSpec);
+
+                    allRuntimeNames.Add(runtimeSpec.Name);
+
+                    // Load dependencies provided by this runtime file
+                    foreach (var dependencySpec in runtimeSpec.Dependencies.Values)
+                    {
+                        // New dependency specs replace existing ones
+                        // This matches the NuGet3 behavior: https://github.com/NuGet/NuGet3/blob/dev/src/NuGet.RuntimeModel/RuntimeDescription.cs#L99
+                        effectiveRuntimeSpecs[dependencySpec.Name] = dependencySpec;
```
Comment (anurse) [inline src/Microsoft.Dnx.Tooling/Restore/RestoreCommand.cs]: After comparing with the NuGet3 code, this is the correct ordering (the original fix did the reverse)

**Comment by anurse** [comment]
Travis failure seems unrelated.

**Comment by anurse** [comment]
ping @davidfowl @dougbu @rynowak

**Comment by davidfowl** [comment]
This is mostly porting the nuget code right?

**Comment by anurse** [comment]
Not quite. The NuGet code has a lot of dependencies (NuGet.RuntimeModel, NuGet.LibraryModel, etc.). But this is basically imitating the way the NuGet code built the list of runtime dependencies.

**Comment by anurse** [comment]
This is the last day of Beta8, can I get a squirrel from someone?

**Comment by davidfowl** [comment]
:shipit:

**Pull request merged**